src/detection_win.cpp

```diff
@@ -328,20 +346,32 @@ void BuildInitialDeviceList() {
 	SP_DEVINFO_DATA* pspDevInfoData = (SP_DEVINFO_DATA*) HeapAlloc(GetProcessHeap(), 0, sizeof(SP_DEVINFO_DATA));
 	pspDevInfoData->cbSize = sizeof(SP_DEVINFO_DATA);
 	for(int i=0; DllSetupDiEnumDeviceInfo(hDevInfo, i, pspDevInfoData); i++) {
-		DWORD nSize=0 ;
-
-		if (!DllSetupDiGetDeviceInstanceId(hDevInfo, pspDevInfoData, buf, sizeof(buf), &nSize)) {
+		if (!DllSetupDiGetDeviceInstanceId(hDevInfo, pspDevInfoData, buf, sizeof(buf), NULL)) {
 			break;
 		}
 
-		DeviceItem_t* item = new DeviceItem_t();
-		item->deviceState = DeviceState_Connect;
-
-		DWORD DataT;
-		DllSetupDiGetDeviceRegistryProperty(hDevInfo, pspDevInfoData, SPDRP_HARDWAREID, &DataT, (PBYTE)buf, MAX_PATH, &nSize);
+		PSP_DEVICE_INTERFACE_DETAIL_DATA pspDeviceIntDetail = NULL;
+		if (!GetDeviceDetails(hDevInfo, pspDevInfoData, i, &pspDeviceIntDetail)) {
+			continue;
+		}
 
-		AddItemToList(buf, item);
-		ExtractDeviceInfo(hDevInfo, pspDevInfoData, buf, MAX_PATH, &item->deviceParams);
+		// Make sure we can actually talk to this device before adding
+		// it to the list of connected devices.
+		HANDLE devHandle = CreateFile(pspDeviceIntDetail->DevicePath,
+									GENERIC_READ | GENERIC_WRITE,
+									FILE_SHARE_READ | FILE_SHARE_WRITE,
+									NULL, OPEN_EXISTING,
+									FILE_FLAG_OVERLAPPED, NULL);
+		if (INVALID_HANDLE_VALUE != devHandle) {
+			CloseHandle(devHandle);
```

Comment:
Seems like we should `CloseHandle(devHandle)` regardless if it is invalid?

Why do we need to talk to the device before adding it to the list?

Reply:
If devHandle is invalid, then there is no handle to close. CloseHandle only needs to be called when CreateFile successfully returns a valid handle.

Talking to the device is another existence verification step before attempting to extract the device information.